### ISSUE TITLE
fileIO primer fails on m1 mac with llvm

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#19218]
`test/release/examples/primers/fileIO.chpl` fails with `EXC_BAD_ACCESS` when using llvm. If I switch `CHPL_TARGET_COMPILER` to clang things seem to work. I also cannot reproduce this on my x86 mac.

```
chpl test/release/examples/primers/fileIO.chpl -g
./fileIO
...
Read line:  of UTF-8 Euro Sign: €
zsh: bus error  ./fileIO
```

```
chpl test/release/examples/primers/fileIO.chpl -g  --target-compiler=clang
./fileIO
...
Read line:  of UTF-8 Euro Sign: €
Running Example 5
Running Example 6
...
Running Example 7
<success>
```

My chplenv:

```
./util/printchplenv --all --anonymize
CHPL_HOST_PLATFORM: darwin
CHPL_HOST_COMPILER: clang
  CHPL_HOST_CC: clang
  CHPL_HOST_CXX: clang++
CHPL_HOST_ARCH: arm64
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: llvm
  CHPL_TARGET_CC: /opt/homebrew/Cellar/llvm@11/11.1.0_4/bin/clang *
  CHPL_TARGET_CXX: /opt/homebrew/Cellar/llvm@11/11.1.0_4/bin/clang++ *
CHPL_TARGET_ARCH: arm64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
  CHPL_GPU_CODEGEN: none
CHPL_COMM: none
CHPL_TASKS: fifo
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_HOST_MEM: cstdlib
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: none
CHPL_RE2: bundled
CHPL_LLVM: system
  CHPL_LLVM_CONFIG: /opt/homebrew/opt/llvm@11/bin/llvm-config
CHPL_AUX_FILESYS: none
CHPL_LIB_PIC: none
CHPL_SANITIZE: none
CHPL_SANITIZE_EXE: none
```

**Comment by daviditen** [comment]
I've simplified the test case down to a more workable case that still trips a bus error on this system.

```chapel
use IO;

record R {
  var i:int;
}

proc file._style2 {
  var ret:iostyleInternal;

  extern proc qio_file_get_style(f:qio_file_ptr_t, ref style:iostyleInternal);
  qio_file_get_style(this._file_internal, ret);

  return ret;
}

proc file.linesHelper2(in local_style:iostyleInternal = this._style2) {
  var ret:R;
  return ret;
}

proc runTest(testfile: string) {
  var f = open(testfile, iomode.cwr);
  f.linesHelper2();
}

runTest("test.bin");
```

**Comment by mppf** [comment]
@daviditen - I think that it is somewhat likely, given that reproducer, that the error has to do with the code to try to generate a call to `qio_file_get_style` in a way that matches the C ABI on ARM.

**Comment by daviditen** [comment]
I simplified the case further to something that no longer uses the `IO` module or the `qio_file_get_style` function. It appears related to passing the `extern iostyleInternal` record to a function by `in` intent. This one actually hits a segfault instead of a bus error. Small differences in the code were switching between the two types of errors.

```chapel
extern record iostyleInternal { }
record R { }

proc getStyle() {
  var ret:iostyleInternal;
  return ret;
}

proc linesHelper(in local_style:iostyleInternal) {
  var ret:R;
  return ret;
}

var s = getStyle();
linesHelper(s);
```

I may have simplified too far - I don't think the record returned by `getStyle()` is initialized so this is probably technically not legal. It's not used anywhere but the assignment to `s` and the copy-in to `linesHelper` though. The crash seems to be happening around the return from `linesHelper`.

**Comment by bradcray** [comment]
>  I don't think the record returned by getStyle() is initialized

I think @mppf recently did a change where Chapel variables of extern type are 0-initialized in https://github.com/chapel-lang/chapel/pull/18542, so you're probably OK in this regard?

**Comment by ronawho** [comment]
Any update on this? It's not blocking me personally but I think more and more of us are using m1's so it'd be nice to take care of any outstanding issues (I _think_ @benharsh ran into a variation of this recently)

**Comment by benharsh** [comment]
I spent some time looking into this issue and I believe it is a genuine LLVM bug that we are encountering.

In the LLVM IR that we generate we are passing the extern struct by value, which for some reason leads to LLVM generating incorrect assembly code for the arguments to ``lineHelper``. I have a pure LLVM IR example in about ~80 lines of code that I would like to use to submit a bug to LLVM when I have the time. I can elaborate more on the exact assembly issues if anyone is interested.

The reason that the clang backend works is that it copies the struct at the callsite and passes it in by reference. I think this is technically a separate ABI compatibility issue.

I think we can resolve both issues by adjusting our LLVM IR to match clang's in this case. More specifically, we should implement the in-intent for extern records by doing a bit-copy at the call-site and passing the result by reference. I am willing to take this on while the issue is still fresh in my mind, and will try to find confirmation that this is indeed matches clang's ABI.

Doing this would sidestep the LLVM IR bug, and allow the fileIO test to finish successfully.

**Comment by bradcray** [comment]
Musing on @benharsh's report this morning, I found myself wondering whether we'd want to / be able to put together a PR that would make M1 Macs default to the C back-end over this.  Or do we think it's sufficiently off in a corner that it likely won't be a problem for users?

**Comment by ronawho** [comment]
Defaulted to C back-end in https://github.com/chapel-lang/chapel/pull/20155.

**Comment by benharsh** [comment]
Resolved by #20270
